awebox/mdl/dynamics_components_dir/mass.py

```diff
@@ -0,0 +1,213 @@
+                                                                 'scaling')
+
+    # this will be used to scale the forces,
+    # so we need to repeat the scaling forces 3x, once per dimension for force.
+    # and, only on those nodes for which we will construct dynamics equations via. Lagrangian mechanics
+
+    node_masses_scaling_stacked = []
+    for node in range(1, number_of_nodes):
+        parent = parent_map[node]
+        mass = node_masses_scaling['m' + str(node) + str(parent)]
+
+        # if node == 1:
+        #     mass += node_masses_scaling['groundstation']
+        #
+
+        node_masses_scaling_stacked = cas.vertcat(node_masses_scaling_stacked, mass, mass, mass)
+
+
+    return node_masses_scaling_stacked
+
+
+def remove_wound_tether_entry(node_masses):
+    if 'm00' in node_masses.keys():
+        node_masses.pop('m00')
+
+    return node_masses
+
+def initialize_mass_dictionary(options, architecture):
+
+    number_of_nodes = architecture.number_of_nodes
+    parent_map = architecture.parent_map
+
+    empty = cas.DM.zeros((1, 1))
+
+    # initialize dictionary
+    node_masses = {}
+    for node in range(1, number_of_nodes):
+        parent = parent_map[node]
+        node_masses['m' + str(node) + str(parent)] = empty
+
+    node_masses['m00'] = empty  # mass added to the ground-station
+    node_masses['groundstation'] = empty
+
+    return node_masses
+
+
+def generate_mass_dictionary_for_all_nodes(options, variables, parameters, architecture, vals_or_scaling):
+
+    number_of_nodes = architecture.number_of_nodes
+    kite_nodes = architecture.kite_nodes
+
+    use_wound_tether = options['tether']['use_wound_tether']
+
+    node_masses = initialize_mass_dictionary(options, architecture)
+
+    upper_nodes_of_above_ground_segments = range(1, number_of_nodes)
+    for upper_node in upper_nodes_of_above_ground_segments:
+        node_masses = add_above_ground_tether_segment_mass(upper_node, node_masses, options,
+                                                           architecture, variables, parameters,
+                                                           vals_or_scaling)
+
+    if use_wound_tether:
+        node_masses = add_wound_tether_mass(node_masses, options, architecture, variables, parameters, vals_or_scaling)
+
+    for kite in kite_nodes:
+        node_masses = add_kite_mass(kite, node_masses, architecture, parameters)
+
+    node_masses = add_groundstation_mass(node_masses, parameters)
+
+    if not use_wound_tether:
+        node_masses = remove_wound_tether_entry(node_masses)
+
+    return node_masses
+
+
+def add_groundstation_mass(node_masses, parameters):
+    m_groundstation = parameters['theta0', 'ground_station', 'm_gen']
+    node_masses['groundstation'] += m_groundstation
+
+    return node_masses
+
+
+def add_wound_tether_mass(node_masses, options, architecture, variables, parameters, vals_or_scaling):
```

Comment:
is this the tether mass that is wound up in the generator s.t. no potential energy just "disappears"?

Reply:
yes. it's a replacement for the mass-correction that was previously being used (though that option is still available)

(well... not so much potential energy (since the ground station is at ground level), but kinetic energy. )